ocaml/lib/snarky_bindings.ml

```diff
@@ -113,43 +115,218 @@ module Field' = struct
   let to_constant_and_terms x = Field.to_constant_and_terms x
 end
 
+let add_gate (label : string) gate =
+  Impl.with_label label (fun () ->
+      Impl.assert_
+        { annotation = None
+        ; basic =
+            Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint.T
+              gate
+        } )
+
 module Gates = struct
+  let zero in1 in2 out =
+    add_gate "zero"
+      (Raw { kind = Zero; values = [| in1; in2; out |]; coeffs = [||] })
+
   let generic sl l sr r so o sm sc =
-    Impl.with_label "generic_gate" (fun () ->
-        Impl.assert_
-          { annotation = Some __LOC__
-          ; basic =
-              Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint.T
-                (Basic { l = (sl, l); r = (sr, r); o = (so, o); m = sm; c = sc })
-          } )
+    add_gate "generic_gate"
+      (Basic { l = (sl, l); r = (sr, r); o = (so, o); m = sm; c = sc })
+
+  let poseidon state = add_gate "poseidon" (Poseidon { state })
+
+  let ec_add p1 p2 p3 inf same_x slope inf_z x21_inv =
+    add_gate "ec_add"
+      (EC_add_complete { p1; p2; p3; inf; same_x; slope; inf_z; x21_inv }) ;
+    (* TODO: do we need this? *)
```

Comment:
do you mean the return value?

Reply:
yeah

Reply:
I think we dont. That was probably my fault when I originally rewrote the `Group.add` function in TypeScript

https://github.com/o1-labs/o1js/blob/main/src/lib/group.ts#L173-L181

can probably just return

```ts
(x3, y3)
```